src/table.js

```diff
@@ -684,10 +760,107 @@ export function __table(source, operations) {
       }))
     );
   }
-  if (primitive) source = source.map((d) => d.value);
   if (source !== input) {
     if (schema) source.schema = schema;
     if (columns) source.columns = columns;
   }
   return source;
 }
+
+function coerceRow(object, types, schema) {
+  const coerced = {};
+  for (const col of schema) {
+    const type = types.get(col.name);
+    const value = object[col.name];
+    coerced[col.name] = type === "raw" ? value : coerceToType(value, type);
+  }
+  return coerced;
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      typeCounts[col][inferType(d[col])]++;
+    }
+  }
+  for (const col in typeCounts) {
+    let type = greatest(Object.keys(typeCounts[col]), d => typeCounts[col][d]);
+    const numDefined = sample.filter(d => !(d[col] == null || d[col] === "")).length;
+    // If over 90% of the sampled data counted as this type, use it. Otherwise,
+    // use "other."
+    type = typeCounts[col][type] / numDefined >= 0.9 ? type : "other";
+    schema.push({
+      name: col,
+      type: type,
+      inferred: type
+    });
+  }
+  return schema;
+}
+
+function inferType(colValue) {
+  const type = typeof colValue;
+  const value = type === "string" ? colValue.trim() : colValue;
+  // for json and sqlite, we already have some types, but for csv and tsv, all
+  // columns are strings here.
+  const typedNonStrings = ["bigint", "boolean", "object"];
+  if (type !== "string") {
+    if (Array.isArray(value)) return "array";
+    else if (value instanceof Date) return "date";
+    else if (value instanceof ArrayBuffer) return "buffer";
+    else if (type === "number") {
+      if (Number.isInteger(+value)) return "integer";
+      else return "number";
+    }
+    else if (typedNonStrings.includes(type)) return type;
+    else return "other";
+  } else {
+    if (value === "true" || value === "false") return "boolean";
```

Comment:
This edge case may not be worth considering, but if you have a mix of `"true"` and `"False"` (note the capitalization), you get an inferred type of "other". Not sure if it's worth also applying `.toLowerCase`?

<img width="928" alt="Screen Shot 2023-02-02 at 11 52 41 AM" src="https://user-images.githubusercontent.com/1373882/216389583-b9d5bce5-75bf-4b52-b188-7498c9c6f33d.png">

Reply:
I expect the inferred type of `"string"` here (or `"boolean"` if we support case-insensitive coercion of `"False"` to `false`, but I don’t think that’s a requirement).

Reply:
I'm in favor of adding case-insensitive boolean inference/coercion, because of this example. The split between inferred booleans and strings means that neither breaks 90%, so we fall back to "other."

Reply:
> The split between inferred booleans and strings means that neither breaks 90%, so we fall back to "other."

I still think `"string"` is more appropriate in this case because it is a superset; the input value `"true"` is both a valid boolean and a valid string.

Reply:
Definitely agree that `"string"` is more appropriate, but we'd quickly get into complexity if we tried adding more interpretation and handling on top of the type count results. In this case we ended up with `{boolean: 5, string: 5}`. (but adding `toLowerCase` does fix that, and seems like a sensible thing to support.)

Reply:
I think we’ll have to solve this problem. There’s a similar issue with integer vs. number. For example, imagine you have the dataset [0.1, 0.2, 0.3, 0.4, 0.5, 1, 2, 3, 4, 5]: half are `"integer"` and half are `"number"`, but neither represents ≥90% and so the inferred type is `"other"`. In this case we should infer the type as `"number"` because an integer is a subtype of number. Here’s a (currently failing) test we can use:

```js
  it("infers mixed integers and numbers as numbers", () => {
    assert.deepStrictEqual(
      inferSchema([0.1, 0.2, 0.3, 0.4, 0.5, 1, 2, 3, 4, 5].map((x) => ({x}))),
      [{name: "x", type: "number", inferred: "number"}]
    );
  });
```

I guess what this means is that we can use the inferType(value) pattern (sorry!) and instead we have to support the idea that values can have multiple possible types. So the type counts for this column would look like `{integer: 5, number: 10}` and the totals of _all_ type counts could be greater than the number of (sampled) values, and we’d still use the greatest type count to infer the type (and return `"other"` if the greatest type count is <90%).

Reply:
Oooh, interesting! Gonna give this a try. 🙏

Reply:
Here’s a test for the true/false/other string case, too (using [PoltiFact’s Truth-o-Meter ratings](https://www.politifact.com/truth-o-meter/)):

```js
  it("infers boolean-ish strings and strings as strings", () => {
    assert.deepStrictEqual(
      inferSchema(["true", "false", "pants on fire"].map((x) => ({x}))),
      [{name: "x", type: "string", inferred: "string"}]
    );
  });
```

Reply:
This is a nice solution, thanks for that, @mbostock. Pushed this update, and seems to work as desired, although one interesting edge case is that `NaN` gets counted as a number, so if you have a mix of integers and `NaN`s, it will infer a type of `"number"` rather then `"integer"`.  Doesn't seem like a show-stopping issue, but thought I'd mention.

![Screen Shot 2023-02-02 at 12 17 51 PM](https://user-images.githubusercontent.com/111310561/216442704-bbda1dee-8a92-46c5-840e-9f90e483cb5e.png)